src/libmongoc/src/mongoc/mongoc-host-list.c

```diff
@@ -343,13 +343,27 @@ _mongoc_host_list_from_hostport_with_err (mongoc_host_list_t *link_,
    if (strchr (host, ':')) {
       link_->family = AF_INET6;
 
-      mongoc_lowercase (link_->host, link_->host);
-      bson_snprintf (link_->host_and_port,
-                     sizeof link_->host_and_port,
-                     "[%s]:%hu",
-                     link_->host,
-                     link_->port);
+      // Check that IPv6 literal is two less than the max to account for `[` and
+      // `]` added below.
+      if (host_len > BSON_HOST_NAME_MAX - 2) {
+         bson_set_error (
+            error,
+            MONGOC_ERROR_STREAM,
+            MONGOC_ERROR_STREAM_NAME_RESOLUTION,
+            "IPv6 literal provided in URI is too long, max is %d chars",
+            BSON_HOST_NAME_MAX - 2);
+         return false;
+      }
 
+      mongoc_lowercase (link_->host, link_->host);
+      int req = bson_snprintf (link_->host_and_port,
+                               sizeof link_->host_and_port,
+                               "[%s]:%hu",
+                               link_->host,
+                               link_->port);
+      BSON_ASSERT (bson_in_range_size_t_signed (req));
+      // Use `<`, not `<=` to account for NULL byte.
+      BSON_ASSERT ((size_t) req < sizeof link_->host_and_port);
```

Comment:
Per snprintf spec:

> The snprintf function returns the number of characters that would have been written had n been sufficiently large, not counting the terminating null character, or a negative value if an encoding error occurred. Thus, the null-terminated output has been completely written if and only if the returned value is nonnegative and less than n.

Has the port length already been validated at this point? If not, this assertion should probably be made a call to `bson_set_error()` instead (e.g. if port ends up being 6 digits long, violating `host_and_port` length expectations).

This comment also applies to the branch below.

Reply:
> Has the port length already been validated at this point?

I expect the port length cannot exceed 5 digits. UINT16_MAX is 65535.